### ISSUE TITLE
Named function/class declarations in default export positions are not mangled

### ISSUE THREAD

**Issue by Andarist** [terser-js/terser#163]
**Bug report or Feature request?**

Bug~

**Version (complete output of `terser -V`)**

latest - 3.10.11 (used with npx)

**Complete CLI command or `minify()` options used**

```
 npx terser -cm --toplevel src/index.js
```

**`terser` input**

```js
export default function loop() {}
```

**`terser` output or error**

```js
export default function loop(){}
```

**Expected result**
```js
export default function(){}
```

**Comment**
Actually they should be removed, splitting the declaration to:
```js
function loop(){}
export default loop
```
removes the name correctly.

Keep in mind that I have given an example using function declaration, but the same thing happens for class declarations too.

**Comment by kzc** [comment]
> terser output or error
>
> export default function loop(){}

This was done intentionally and will not change. 

`terser` and `uglify-es` before it will not alter or drop the names of ES6 exports - even default export. There were too many problems in the wild otherwise. Rollup's source code was used as a test case for `uglify-es` back in the day and dropping the name yielded failing results when minifying its code base in-situ and running "npm test".

> latest (used with npx)

Please fill in the version so that future searches will yield useful results.

**Comment by Andarist** [comment]
> This was done intentionally and will not change.

That's quite weird - named exports can be mangled (and rollup does it when doing its codesplitting!). Default exports are not accessible by their names so Im not sure whats the technical reason behind keeping them? IMHO they should only be kept when using `keep_fnames` option.

> Please fill in the version so that future searches will yield useful results.

Done

**Comment by kzc** [comment]
> Bug report or Feature request?
>
> Bug~

Just to clarify, this is a feature request. The minified code runs correctly as per the issue template.

```
<!-- Note: sub-optimal but correct code is not a bug -->
```

> named exports can be mangled (and rollup does it when doing its codesplitting!) by mangling local identifiers

As does Terser:
```
$ echo 'function loop() {} export {loop as run};' | terser --module -mc
function n(){}export{n as run};
```

I remember one of the issues with `export default` now. The proposed change of `export default function loop(){}` to `export default function(){}` would change the function definition `AST_Defun` to a function expression `AST_Function`. That created problems as I recall.

I'd be willing to review a PR implementing a feature behind a new `mangle` option named `default_exports` defaulting to `false`. If this option were to be set to `true` then `minify` could mangle (but retain) the export default function definition name.

**Comment by fabiosantoscode** [comment]
I'd posit that we don't need an option to remove export names. The default works, so it should be used. It's only a few bytes anyway

**Comment by Andarist** [comment]
> I remember one of the issues with export default now. The proposed change of export default function loop(){} to export default function(){} would change the function definition AST_Defun to a function expression AST_Function. That created problems as I recall.

That seems like an uglify/terser implementation problem, not a semantic problem that cant be overcome.

> The default works, so it should be used. It's only a few bytes anyway
> Yeah, the current behavior is fine. Might as well close this issue then.

I'm slightly surprised by this judgement. This is a minifier project and should strive for the best results - especially that from the point of static analysis this one is **easy** to do. I'm not suggesting that this has to changed right away or that you should definitely waste your time resources on this one, it's only a thing I have noticed and I personally consider a bug (because it doesn't follow what mangle option should do by default without a semantic argument behind it), so I've created an issue about it.

To sum it up - I don't care about this issue being resolved any time soon (or any time at all) as bundlers make this issue non-existent for me.

That being said it IMHO should be considered an issue (considering an open issue as a task in a backlog) that should at least stay open until it gets resolved.

**Comment by Andarist** [comment]
> @Andarist Being lectured by another open source contributor about "best results" and what constitutes a bug is frankly insulting - particularly in light of how many hundreds of unpaid hours I put into this project getting it to its present level of functionality and reliability.

If my comments have sounded harsh, pretentious or impolite - I'm very sorry. I have uttermost for you and the work you have done and I don't expect you do anything if you don't want to - I understand how hard & exhausting maintaining the project is, not even mentioning developing it. My comments were meant purely as part of the discussion. English is not my native language so it's harder for me to structure sentences in a manner that emphasizes my positive emotions associated with stating my opinions etc.

That being said I personally still would classify it more of a low priority bug (which is not causing problems in this case). If I'd report something like `function test(){}` keeping its name I suspect you'd classify it as a bug and I don't see how those cases differ (from semantic point of view). I respect your other opinion on this though, also in the end it doesn't really matter in this case.

Once again - I apologize if I have sounded like I was lecturing you, I have not intended to do so.

OSS ❤️

**Comment by kzc** [comment]
@Andarist No worries. I misinterpreted the tone of the comment. It wouldn't be the first time.

> That being said I personally still would classify it more of a low priority bug (which is not causing problems in this case).

What case wouldn't work? All programmers are pedantic - particularly ones that work on minifiers. I thought this statement in the issue template was unambiguous:

> `<!-- Note: sub-optimal but correct code is not a bug -->`

Not sure how it could be more clear. Must the word "correct" be replaced with "correctly functioning"? No matter.

I went back to the Rollup test suite and am pleasantly surprised to report that it runs successfully when minification was applied with unnamed `export default` functions. So something must have been fixed in one of the projects since I last looked at this issue.

I now see the use case to mangle `export default` function (or class) definition names in cases where it is referenced within the module. If I have some time I can look at implementing both of these heuristics - definition to expression conversion if safe, otherwise mangle. 

If you put a PR together I'd be happy to review it.

This is what Terser AST looks like:

```
$ echo 'export default function FuncName(){}; FuncName();' | terser -mc --module -o ast
{
  "_class": "AST_Toplevel",
  "variables": [
    "1000004 FuncName"
  ],
  "functions": [
    "1000004 FuncName"
  ],
  "enclosed": [
    "1000004 FuncName"
  ],
  "body": [
    {
      "_class": "AST_Export",
      "exported_definition": {
        "_class": "AST_Defun",
        "name": {
          "_class": "AST_SymbolDefun",
          "name": "FuncName",
          "thedef": "1000004 FuncName"
        },
        "argnames": [],
        "uses_arguments": false,
        "is_generator": false,
        "async": false,
        "variables": [
          "1000003 arguments"
        ],
        "body": []
      },
      "is_default": true
    },
    {
      "_class": "AST_SimpleStatement",
      "body": {
        "_class": "AST_Call",
        "expression": {
          "_class": "AST_SymbolRef",
          "name": "FuncName",
          "thedef": "1000004 FuncName"
        },
        "args": []
      }
    }
  ]
}
```
The AST class definitions are in `lib/ast.js`. A summary of AST classes can be seen with `terser -h ast`. The feature would be implemented in `lib/scope.js` (mangle) and `lib/compress.js` (compress).
